File: solutions/clojure/isbn-verifier/2/src/isbn_verifier.clj
(ns isbn-verifier
  (:require [clojure.string :as str]))

;(x1 * 10 + x2 * 9 + x3 * 8 + x4 * 7 + x5 * 6 + x6 * 5 + x7 * 4 + x8 * 3 + x9 * 2 + x10 * 1) mod 11 == 0

(defn digit->int
  [c]
  (cond-> c
    (char? c) (Character/digit 10)))


(defn valid?
  [parsed]
  (->> parsed
       (replace {\X 10})
       (map digit->int)
       (map * (range 10 0 -1))
       (apply +)
       (#(mod % 11))
       zero?))


(def isbn-regex #"(\d)-?(\d{3})-?(\d{5})-?([\dX])")

(defn isbn?
  [isbn]
  (true?
   (when-let [parsed (re-matches isbn-regex isbn)] 
     (->> parsed
          (drop 1)
          (apply str)
          valid?))))

File: solutions/java/isbn-verifier/2/src/main/java/IsbnVerifier.java
import java.util.regex.Pattern;
import java.util.stream.IntStream;

class IsbnVerifier {
    private static final Pattern ISBN_PATTERN = Pattern.compile("\\d{9}[\\dX]");

    boolean isValid(String stringToVerify) {
        var isbn = stringToVerify.replace("-", "");

        return ISBN_PATTERN.matcher(isbn).matches() && checksum(isbn) % 11 == 0;
    }

    private static int checksum(String isbn) {
        return IntStream.range(0, isbn.length())
                .map(index -> digitValue(isbn.charAt(index)) * (10 - index))
                .sum();
    }

    private static int digitValue(char character) {
        return character == 'X' ? 10 : Character.digit(character, 10);
    }
}
